File: src/core/route.ts
// Straight and single-elbow line routing, plus endpoint connection and
// orientation inference for the line/arrow tools.
// Ported from ASCIIFlow (client/draw/utils.ts, client/draw/line.ts), MIT © Lewis Hemens.
import { connect, connectable, connects, disconnect, isSpecial, UNICODE } from "./glyphs";
import { Layer, StackedLayers, type LayerView } from "./layer";
import { Direction, Vector } from "./vector";

export function line(start: Vector, end: Vector, horizontalFirst: boolean): Layer {
  if (start.x === end.x || start.y === end.y) return straightLine(start, end);
  return cornerLine(start, end, horizontalFirst);
}

export function cornerLine(start: Vector, end: Vector, horizontalFirst: boolean): Layer {
  const corner = horizontalFirst ? new Vector(end.x, start.y) : new Vector(start.x, end.y);
  const layer = straightLine(start, corner);
  layer.setFrom(straightLine(corner, end));
  const right = start.x < end.x;
  const down = start.y < end.y;
  let glyph: string;
  if (horizontalFirst) {
    glyph = right
      ? down
        ? UNICODE.cornerTopRight
        : UNICODE.cornerBottomRight
      : down
        ? UNICODE.cornerTopLeft
        : UNICODE.cornerBottomLeft;
  } else {
    glyph = down
      ? right
        ? UNICODE.cornerBottomLeft
        : UNICODE.cornerBottomRight
      : right
        ? UNICODE.cornerTopLeft
        : UNICODE.cornerTopRight;
  }
  layer.set(corner, glyph);
  return layer;
}

export function straightLine(start: Vector, end: Vector): Layer {
  const layer = new Layer();
  if (start.x !== end.x && start.y !== end.y) {
    throw new Error(`Can't draw a straight line between ${start} and ${end}`);
  }
  if (start.x === end.x) {
    const top = Math.min(start.y, end.y);
    const bottom = Math.max(start.y, end.y);
    for (let y = top; y <= bottom; y++) layer.set(new Vector(start.x, y), UNICODE.lineVertical);
  }
  if (start.y === end.y) {
    const left = Math.min(start.x, end.x);
    const right = Math.max(start.x, end.x);
    for (let x = left; x <= right; x++) layer.set(new Vector(x, start.y), UNICODE.lineHorizontal);
  }
  return layer;
}

/** Arrow head glyph at `end` for a route from `start`. */
export function arrowFor(start: Vector, end: Vector, horizontalFirst: boolean): string {
  if (end.x === start.x) return end.y < start.y ? UNICODE.arrowUp : UNICODE.arrowDown;
  if (end.y === start.y) return end.x < start.x ? UNICODE.arrowLeft : UNICODE.arrowRight;
  if (horizontalFirst) return end.y < start.y ? UNICODE.arrowUp : UNICODE.arrowDown;
  return end.x > start.x ? UNICODE.arrowRight : UNICODE.arrowLeft;
}

interface CellContext {
  left: boolean;
  right: boolean;
  up: boolean;
  down: boolean;
  leftup: boolean;
  leftdown: boolean;
  rightup: boolean;
  rightdown: boolean;
}

export function cellContext(p: Vector, layer: LayerView): CellContext {
  const s = (v: Vector) => isSpecial(layer.get(v));
  return {
    left: s(p.left()),
    right: s(p.right()),
    up: s(p.up()),
    down: s(p.down()),
    leftup: s(p.left().up()),
    leftdown: s(p.left().down()),
    rightup: s(p.right().up()),
    rightdown: s(p.right().down()),
  };
}

/**
 * Elbow orientation for a drag from `start` to `end`, inferred from the
 * structure around both endpoints. `flip` reverses the inference.
 */
export function inferHorizontalFirst(start: Vector, end: Vector, committed: LayerView, flip: boolean): boolean {
  const s = cellContext(start, committed);
  const e = cellContext(end, committed);
  const horizontalStart = (s.up && s.down) || (s.leftup && s.leftdown) || (s.rightup && s.rightdown);
  const verticalEnd = (e.left && e.right) || (e.leftup && e.rightup) || (e.leftdown && e.rightdown);
  return (horizontalStart || verticalEnd) !== flip;
}

/**
 * Connects the given endpoint cells of `layer` to any structure pointing at
 * them, then trims connections that lead nowhere.
 */
export function connectEndpoints(layer: Layer, base: Layer, ends: readonly Vector[]): void {
  const combined = new StackedLayers([base, layer]);
  for (const p of ends) {
    const incoming = Direction.ALL.filter(
      (d) => connects(combined.get(p.add(d)), d.opposite()) && connectable(layer.get(p), d),
    );
    layer.set(p, connect(layer.get(p), incoming)!);
    layer.set(p, disconnect(layer.get(p), Direction.ALL.filter((d) => !incoming.includes(d)))!);
  }
}

File: src/core/route.rs
//! Straight and single-elbow line routing, plus endpoint connection and
//! orientation inference for the line/arrow tools.
//!
//! Ported from ASCIIFlow (`client/draw/utils.ts`, `client/draw/line.ts`), MIT © Lewis Hemens.

use super::glyphs::{connect_all, connectable, connects, disconnect_all, is_special, UNICODE};
use super::layer::Layer;
use super::vector::{Direction, Pos};

pub fn line(start: Pos, end: Pos, horizontal_first: bool) -> Layer {
    if start.x == end.x || start.y == end.y {
        return straight_line(start, end);
    }
    corner_line(start, end, horizontal_first)
}

pub fn corner_line(start: Pos, end: Pos, horizontal_first: bool) -> Layer {
    let corner = if horizontal_first {
        Pos::new(end.x, start.y)
    } else {
        Pos::new(start.x, end.y)
    };
    let mut layer = straight_line(start, corner);
    layer.set_from(&straight_line(corner, end));
    let right = start.x < end.x;
    let down = start.y < end.y;
    let glyph = if horizontal_first {
        match (right, down) {
            (true, true) => UNICODE.corner_top_right,
            (true, false) => UNICODE.corner_bottom_right,
            (false, true) => UNICODE.corner_top_left,
            (false, false) => UNICODE.corner_bottom_left,
        }
    } else {
        match (down, right) {
            (true, true) => UNICODE.corner_bottom_left,
            (true, false) => UNICODE.corner_bottom_right,
            (false, true) => UNICODE.corner_top_left,
            (false, false) => UNICODE.corner_top_right,
        }
    };
    layer.set(corner, glyph);
    layer
}

pub fn straight_line(start: Pos, end: Pos) -> Layer {
    let mut layer = Layer::new();
    assert!(
        start.x == end.x || start.y == end.y,
        "can't draw a straight line between {start} and {end}"
    );
    if start.x == end.x {
        let (top, bottom) = (start.y.min(end.y), start.y.max(end.y));
        for y in top..=bottom {
            layer.set(Pos::new(start.x, y), UNICODE.line_vertical);
        }
    }
    if start.y == end.y {
        let (left, right) = (start.x.min(end.x), start.x.max(end.x));
        for x in left..=right {
            layer.set(Pos::new(x, start.y), UNICODE.line_horizontal);
        }
    }
    layer
}

/// Arrow head glyph at `end` for a route from `start`.
pub fn arrow_head(start: Pos, end: Pos, horizontal_first: bool) -> char {
    if end.x == start.x {
        return if end.y < start.y {
            UNICODE.arrow_up
        } else {
            UNICODE.arrow_down
        };
    }
    if end.y == start.y {
        return if end.x < start.x {
            UNICODE.arrow_left
        } else {
            UNICODE.arrow_right
        };
    }
    if horizontal_first {
        if end.y < start.y {
            UNICODE.arrow_up
        } else {
            UNICODE.arrow_down
        }
    } else if end.x > start.x {
        UNICODE.arrow_right
    } else {
        UNICODE.arrow_left
    }
}

/// Structure around a cell, as eight booleans.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct CellContext {
    pub left: bool,
    pub right: bool,
    pub up: bool,
    pub down: bool,
    pub left_up: bool,
    pub left_down: bool,
    pub right_up: bool,
    pub right_down: bool,
}

pub fn cell_context(p: Pos, layer: &Layer) -> CellContext {
    let s = |v: Pos| layer.get(v).is_some_and(is_special);
    CellContext {
        left: s(p.left()),
        right: s(p.right()),
        up: s(p.up()),
        down: s(p.down()),
        left_up: s(p.left().up()),
        left_down: s(p.left().down()),
        right_up: s(p.right().up()),
        right_down: s(p.right().down()),
    }
}

/// Elbow orientation for a drag from `start` to `end`, inferred from the
/// structure around both endpoints. `flip` reverses the inference.
pub fn infer_horizontal_first(start: Pos, end: Pos, committed: &Layer, flip: bool) -> bool {
    let s = cell_context(start, committed);
    let e = cell_context(end, committed);
    let horizontal_start =
        (s.up && s.down) || (s.left_up && s.left_down) || (s.right_up && s.right_down);
    let vertical_end =
        (e.left && e.right) || (e.left_up && e.right_up) || (e.left_down && e.right_down);
    (horizontal_start || vertical_end) != flip
}

fn combined_get(layer: &Layer, base: &Layer, p: Pos) -> Option<char> {
    layer.get(p).or_else(|| base.get(p))
}

/// Connects the given endpoint cells of `layer` to any structure pointing at
/// them, then trims connections that lead nowhere.
pub fn connect_endpoints(layer: &mut Layer, base: &Layer, ends: &[Pos]) {
    for &p in ends {
        let incoming: Vec<Direction> = Direction::ALL
            .into_iter()
            .filter(|d| {
                let neighbour = combined_get(layer, base, p.add(d.delta()));
                let points_back = neighbour.is_some_and(|n| connects(n, d.opposite()));
                points_back && layer.get(p).is_some_and(|v| connectable(v, *d))
            })
            .collect();
        if let Some(v) = layer.get(p) {
            let connected = connect_all(v, &incoming);
            let rest: Vec<Direction> = Direction::ALL
                .into_iter()
                .filter(|d| !incoming.contains(d))
                .collect();
            layer.set(p, disconnect_all(connected, &rest));
        }
    }
}
